### ISSUE TITLE
Alignment data in Milestones

### ISSUE THREAD

**Issue by jag3773** [ubsicap/usfm#59]
Perhaps related to #58 .

We are hoping to be able to encode alignment data in USFM3.  The word attributes go a long way toward providing this capability, but using them alone does not allow for complex alignments, even including multiple target words to align with a single source word doesn't work.

Instead, milestones, with their ability to overlap, provide an ideal solution to encoding complex alignments.

In the following example we abuse the the [Keyword / keyterm](https://ubsicap.github.io/usfm/characters/index.html#k-k) marker for our milestones because we can't find any that are more appropriate in the USFM 3 specification.  An implementation of #58 that includes support for `-s` and `-e` may suffice for this purpose.

```
\v 1 \k-s | strong="G39720" x-occurrence="1" x-occurrences="1" x-ugnt="Παῦλος"
\w Paul|x-occurrence="1" x-occurrences="1"\w*
\k-e\*,\k-s | strong="G14010" x-occurrence="1" x-occurrences="1" x-ugnt="δοῦλος"
\w a|x-occurrence="1" x-occurrences="1"\w*
\w servant|x-occurrence="1" x-occurrences="1"\w*
\k-e\*
\k-s | strong="G23160" x-occurrence="1" x-occurrences="2" x-ugnt="Θεοῦ"
\w of|x-occurrence="1" x-occurrences="4"\w*
\w God|x-occurrence="1" x-occurrences="2"\w*
\k-e\*
\k-s | strong="G06520" x-occurrence="1" x-occurrences="1" x-ugnt="ἀπόστολος"
\w and|x-occurrence="1" x-occurrences="2"\w*
\w an|x-occurrence="1" x-occurrences="4"\w*
\w apostle|x-occurrence="1" x-occurrences="1"\w*
\k-e\*
\k-s | strong="G11610" x-occurrence="1" x-occurrences="1" x-ugnt="δὲ"
\w of|x-occurrence="2" x-occurrences="4"\w*
\k-e\*
\k-s | strong="G24240" x-occurrence="1" x-occurrences="1" x-ugnt="Ἰησοῦ"
\w Jesus|x-occurrence="1" x-occurrences="1"\w*
\k-e\*
\k-s | strong="G55470" x-occurrence="1" x-occurrences="1" x-ugnt="Χριστοῦ"
\w Christ|x-occurrence="1" x-occurrences="1"\w*
\k-e\*,\k-s | strong="G25960" x-occurrence="1" x-occurrences="1" x-ugnt="κατὰ"
\w for|x-occurrence="1" x-occurrences="1"\w*
\k-e\*
\k-s | strong="G41020" x-occurrence="1" x-occurrences="1" x-ugnt="πίστιν"
\w the|x-occurrence="1" x-occurrences="3"\w*
\w faith|x-occurrence="1" x-occurrences="1"\w*
\k-e\*
\k-s | strong="G23160" x-occurrence="2" x-occurrences="2" x-ugnt="Θεοῦ"
\w of|x-occurrence="3" x-occurrences="4"\w*
\w God's|x-occurrence="2" x-occurrences="2"\w*
\k-e\*
\k-s | strong="G15880" x-occurrence="1" x-occurrences="1" x-ugnt="ἐκλεκτῶν"
\w chosen|x-occurrence="1" x-occurrences="1"\w*
\w people|x-occurrence="1" x-occurrences="1"\w*
\k-e\*
\k-s | strong="G25320" x-occurrence="1" x-occurrences="1" x-ugnt="καὶ"
\w and|x-occurrence="2" x-occurrences="2"\w*
\k-e\*
\k-s | strong="G02250" x-occurrence="1" x-occurrences="1" x-ugnt="ἀληθείας"
\w the|x-occurrence="2" x-occurrences="3"\w*
\k-e\*
\k-s | strong="G19220" x-occurrence="1" x-occurrences="1" x-ugnt="ἐπίγνωσιν"
\w knowledge|x-occurrence="1" x-occurrences="1"\w*
\k-e\*
\k-s | strong="G02250" x-occurrence="1" x-occurrences="1" x-ugnt="ἀληθείας"
\w of|x-occurrence="4" x-occurrences="4"\w*
\w truth|x-occurrence="1" x-occurrences="1"\w*
\w the|x-occurrence="3" x-occurrences="3"\w*
\k-e\*
\k-s | strong="G35880" x-occurrence="1" x-occurrences="1" x-ugnt="τῆς"
\w that|x-occurrence="1" x-occurrences="1"\w*
\k-e\*
\k-s | strong="G25960" x-occurrence="1" x-occurrences="1" x-ugnt="κατ’"
\w agrees|x-occurrence="1" x-occurrences="1"\w*
\w with|x-occurrence="1" x-occurrences="1"\w*
\k-e\*
\k-s | strong="G21500" x-occurrence="1" x-occurrences="1" x-ugnt="εὐσέβειαν"
\w godliness|x-occurrence="1" x-occurrences="1"\w*
\k-e\*,
```

**Comment by jag3773** [comment]
@klassenjm Any ideas on this?  Should we perhaps proceed with using `\mg` as purposed in https://github.com/ubsicap/usfm/issues/58 ?

**Comment by klassenjm** [comment]
@jag3773 . Hi Jesse. Sorry you did not get a reply. Thank you for following up.

I'd like to make it clearer in the documentation that the milestone syntax is a generic syntax, and that like word level attributes there are some types which have been encoded in the standard (e.g. for \w ...\w). I think that when USFM 3.0 is supported will enough, we can discover what other common use of attributes or milestones occur, and offically encode them in 3.1

That's what I'm thinking.

For word-level attributes, you can create user-defined attributes using ``x-`` as a prefix.

Would you be OK with the same pattern for user generated milestones? So, ``\x-k\*`` (to use your k example), or whatever ``x-abcd\*``. I would say that the specification will express that start and end parameters are optional.

Thoughts?

**Comment by jag3773** [comment]
@klassenjm That sounds good to me.  To be clear that I'm understanding, does the following example look acceptable?


```
\v 1 \x-aln-s | strong="G39720" x-occurrence="1" x-occurrences="1" x-ugnt="Παῦλος"
\w Paul|x-occurrence="1" x-occurrences="1"\w*
\x-aln-e\*,\x-aln-s | strong="G14010" x-occurrence="1" x-occurrences="1" x-ugnt="δοῦλος"
\w a|x-occurrence="1" x-occurrences="1"\w*
\w servant|x-occurrence="1" x-occurrences="1"\w*
\x-aln-e\*
```

**Comment by klassenjm** [comment]
@jag3773 This looks to be following the syntax for (user-defined) attributes and milestones correctly.

**Comment by RobH123** [comment]
I can't see what is defining the end of the \x-aln-s milestone there?

BTW, do the Quotation start / end milestones replace the USFM2 \sp marker? (It's not mentioned on the milestones page. And I can't see a list of deprecated/obsolete USFM2 markers in the USFM3 documentation.)

**Comment by klassenjm** [comment]
@RobH123 There is a list of USFM changes here: http://ubsicap.github.io/usfm/about/releasenotes.html which includes deprecated markup in the Marker Revisions section.

There will be a few changes still, because (for example) in implementing Ruby markup support in an editor, requirements were better understood.

In the example from @jag3773, the corresponding end for `\x-aln-s` is `\x-aln-e\*`

This does not replace `\qt ...\qt*`, since character level spans cannot cross a paragraph boundary. That's not visible in Jesse's example - but it is the issue for what milestone were primarily aimed at solving. Jesse is looking for a solution for wrapping short phases of words with metadata.

@jag3773 I realized that the exampled you shared in not quite correct. Any milestone marker has a syntax of `\marker-s|e\*` whether it is a start or end (and the start / end being optional). I also realized my suggestion of `x-` is not ideal in this case. Promoting `x-` for user defined attributes is good, since it's relatively common practice. However, USFM has promoted the `\z` namespace for user defined paragraph or character markup additions. I think we should stick with that, since these fall into the bucket of "things that start with a backslash".

Suggested alternate:

```
\v 1 \zaln-s | strong="G39720" x-occurrence="1" x-occurrences="1" x-ugnt="Παῦλος"\*
\w Paul|x-occurrence="1" x-occurrences="1"\w*
\zaln-e\*,\zaln-s | strong="G14010" x-occurrence="1" x-occurrences="1" x-ugnt="δοῦλος"\*
\w a|x-occurrence="1" x-occurrences="1"\w*
\w servant|x-occurrence="1" x-occurrences="1"\w*
\zaln-e\*
```

USFM 3.0 will only officially recognize `strong` applied to `\w...|...\w*`, so technically speaking I think it needs to be x-strong elsewhere.
